File: src/main/java/util/extension/RobotExtension.java
package util.extension;

import it.unimi.dsi.fastutil.ints.IntArrayList;
import java.awt.*;
import java.awt.event.InputEvent;
import java.awt.event.KeyEvent;
import java.util.HashMap;
import java.util.Map;

import kotlin.Pair;
import lombok.experimental.ExtensionMethod;
import lombok.experimental.UtilityClass;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.lang3.function.TriConsumer;

@Log4j2
@UtilityClass
@ExtensionMethod(MapExtension.class)
public class RobotExtension {

	/**
	 * Width of the screen in pixels.
	 */
	public static final int screenWidth = Toolkit.getDefaultToolkit().getScreenSize().width;

	/**
	 * Height of the screen in pixels.
	 */
	public static final int screenHeight = Toolkit.getDefaultToolkit().getScreenSize().height;

	/**
	 * Maps special characters to sequences of key presses and releases.
	 */
	private static final Map<Pair<String, Character>, IntArrayList> special = new HashMap<>();

	/**
	 * Maps direction names to lambda expressions defining mouse movement actions.
	 */
	private static final Map<String, TriConsumer<Integer, Integer, Robot>> directionActions = new HashMap<>();
	
	static {
		initializeDirections();
		initializeSpecialChars();
	}

	private static void initializeDirections() {
		directionActions.put("up", (x, y, robot) -> robot.mouseMove(x, y - 50));
		directionActions.put("down", (x, y, robot) -> robot.mouseMove(x, y + 50));
		directionActions.put("left", (x, y, robot) -> robot.mouseMove(x - 50, y));
		directionActions.put("right", (x, y, robot) -> robot.mouseMove(x + 50, y));
	}

	private static void initializeSpecialChars() {
		special.put(new Pair<>("question", '?'), IntArrayList.of(KeyEvent.VK_SHIFT, KeyEvent.VK_SLASH));
		special.put(new Pair<>("exclamation", '!'), IntArrayList.of(KeyEvent.VK_SHIFT, KeyEvent.VK_1));
		special.put(new Pair<>("colon", ':'), IntArrayList.of(KeyEvent.VK_SHIFT, KeyEvent.VK_SEMICOLON));
		special.put(new Pair<>("quote", '"'), IntArrayList.of(KeyEvent.VK_SHIFT, KeyEvent.VK_QUOTE));
		special.put(new Pair<>("tilde", '~'), IntArrayList.of(KeyEvent.VK_SHIFT, KeyEvent.VK_BACK_QUOTE));
	}

	/**
	 * Simulates typing a sequence of characters using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @param keys  The sequence of characters to type.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot type(Robot robot, String keys) {
		for (char c : keys.toCharArray()) {
			int keyCode = KeyEvent.getExtendedKeyCodeForChar(c);
			if (KeyEvent.CHAR_UNDEFINED == keyCode) {
				throw new IllegalArgumentException(
						"Key code not found for character '" + c + "'");
      		} else if (special.containsKeySecond(c)) {
				for (int i = 0; i < special.getSecond(c).size(); i++) {
					robot.keyPress(special.getSecond(c).getInt(i));
				}
				for (int i = special.getSecond(c).size() - 1; i >= 0; i--) {
					robot.keyRelease(special.getSecond(c).getInt(i));
				}
			} else if (Character.isUpperCase(c)) {
				shift(robot, keyCode);
			} else {
				type(robot, keyCode);
			}
		}
		return robot;
	}

	/**
	 * Simulates pressing and releasing a single key using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @param i     The virtual key code representing the key to press and release.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot type(Robot robot, int i) {
		robot.keyPress(i);
		robot.keyRelease(i);
		return robot;
	}

	/**
	 * Simulates a left mouse click using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot leftClick(Robot robot) {
		robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
		robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
		return robot;
	}

	/**
	 * Simulates a right mouse click using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot rightClick(Robot robot) {
		robot.mousePress(InputEvent.BUTTON2_DOWN_MASK);
		robot.mouseRelease(InputEvent.BUTTON2_DOWN_MASK);
		return robot;
	}

	/**
	 * Simulates pressing the ENTER key using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot enter(Robot robot) {
		type(robot, KeyEvent.VK_ENTER);
		return robot;
	}

	/**
	 * Simulates pressing the TAB key using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot tab(Robot robot) {
		type(robot, KeyEvent.VK_TAB);
		return robot;
	}

	/**
	 * Simulates pressing the CONTROL key followed by another key using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @param i     The virtual key code representing the key to press after CONTROL.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot control(Robot robot, int i) {
		robot.keyPress(KeyEvent.VK_CONTROL);
		type(robot, i);
		robot.keyRelease(KeyEvent.VK_CONTROL);
		return robot;
	}

	/**
	 * Simulates pressing the SHIFT key followed by another key using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @param i     The virtual key code representing the key to press after SHIFT.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot shift(Robot robot, int i) {
		robot.keyPress(KeyEvent.VK_SHIFT);
		type(robot, i);
		robot.keyRelease(KeyEvent.VK_SHIFT);
		return robot;
	}

	/**
	 * Simulates saving something using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot save(Robot robot) {
		control(robot, KeyEvent.VK_S);
		return robot;
	}

	/**
	 * Moves the mouse cursor according to the specified directions using a {@link Robot}.
	 *
	 * @param robot The {@link Robot} instance to use for simulation.
	 * @param direction A space-separated string specifying the directions ("up", "down", "left", "right").
	 * @return The modified {@link Robot} instance.
	 */
	public static Robot mouseMoveString(Robot robot, String direction) {
		String[] split = direction.split(" ");
		for (String dir : split) {
			int x = MouseInfo.getPointerInfo().getLocation().x;
			int y = MouseInfo.getPointerInfo().getLocation().y;
			TriConsumer<Integer, Integer, Robot> action = directionActions.get(dir);
			if (action != null) {
				action.accept(x, y, robot);
			} else {
              log.debug("Invalid direction: \"{}\"", dir);
			}
		}
		return robot;
	}
}

File: src/main/kotlin/util/extension/RobotExtension.kt
package util.extension

import co.touchlab.kermit.Logger
import util.emu.Direction.Companion.fromString
import util.extension.RobotUtils.special
import java.awt.MouseInfo.getPointerInfo
import java.awt.Robot
import java.awt.event.InputEvent.BUTTON1_DOWN_MASK
import java.awt.event.InputEvent.BUTTON3_DOWN_MASK
import java.awt.event.KeyEvent
import java.awt.event.KeyEvent.CHAR_UNDEFINED
import java.awt.event.KeyEvent.VK_CONTROL
import java.awt.event.KeyEvent.VK_ENTER
import java.awt.event.KeyEvent.VK_F1
import java.awt.event.KeyEvent.VK_F10
import java.awt.event.KeyEvent.VK_F11
import java.awt.event.KeyEvent.VK_F12
import java.awt.event.KeyEvent.VK_F2
import java.awt.event.KeyEvent.VK_F3
import java.awt.event.KeyEvent.VK_F4
import java.awt.event.KeyEvent.VK_F5
import java.awt.event.KeyEvent.VK_F6
import java.awt.event.KeyEvent.VK_F7
import java.awt.event.KeyEvent.VK_F8
import java.awt.event.KeyEvent.VK_F9
import java.awt.event.KeyEvent.VK_S
import java.awt.event.KeyEvent.VK_TAB

/**
 * Moves the mouse cursor based on the specified directions.
 *
 * @param direction A string containing one or more directions (e.g., "up left") to move the mouse
 *   cursor. Each direction should be separated by a space.
 */
fun Robot.mouseMoveString(direction: String) {
    direction.split(" ").mapNotNull { fromString(it) }.forEach { it.tri(getPointerInfo().location.x, getPointerInfo().location.y, this) }
}

/**
 * Simulates typing a sequence of characters.
 *
 * @param keys A string where each character represents a key to be typed by the robot.
 */
fun Robot.type(keys: String) {
    keys.forEach { c ->
        when {
            c == CHAR_UNDEFINED -> Logger.e("Warning: Key code not found for character '\$c'")
            special.containsKeySecond(c) -> special.getFromSecond(c).forEach { key -> type(key) }
            c.isUpperCase() -> shift(KeyEvent.getExtendedKeyCodeForChar(c.code))
            else -> type(KeyEvent.getExtendedKeyCodeForChar(c.code))
        }
    }
}

/**
 * Presses and releases a specified key.
 *
 * @param keyCode The integer code of the key to be pressed and released.
 */
fun Robot.type(keyCode: Int) {
    keyPress(keyCode)
    keyRelease(keyCode)
}

/** Simulates a left mouse click. */
fun Robot.leftClick() = click(BUTTON1_DOWN_MASK)

/** Simulates a right mouse click. */
fun Robot.rightClick() = click(BUTTON3_DOWN_MASK)

/** Simulates pressing the ENTER key. */
fun Robot.enter() = type(VK_ENTER)

/** Simulates pressing the TAB key. */
fun Robot.tab() = type(VK_TAB)

/**
 * Simulates pressing CONTROL + another key.
 *
 * @param keyCode The integer code of the key to be pressed in combination with the CONTROL key.
 */
fun Robot.control(keyCode: Int) {
    keyPress(VK_CONTROL)
    type(keyCode)
    keyRelease(VK_CONTROL)
}

/**
 * Simulates pressing CONTROL + doing something else.
 *
 * @param action The function to be done in combination with the CONTROL key.
 */
fun Robot.control(action: (Robot) -> Unit) {
    keyPress(VK_CONTROL)
    action(this)
    keyRelease(VK_CONTROL)
}

fun Robot.f(i: Int?) {
    if (i == null) return
    when (i) {
        1 -> type(VK_F1)
        2 -> type(VK_F2)
        3 -> type(VK_F3)
        4 -> type(VK_F4)
        5 -> type(VK_F5)
        6 -> type(VK_F6)
        7 -> type(VK_F7)
        8 -> type(VK_F8)
        9 -> type(VK_F9)
        10 -> type(VK_F10)
        11 -> type(VK_F11)
        12 -> type(VK_F12)
    }
}

/**
 * Simulates pressing COMMAND + another key.
 *
 * @param keyCode The integer code of the key to be pressed in combination with the COMMAND key.
 */
fun Robot.command(keyCode: Int) {
    keyPress(KeyEvent.VK_META)
    type(keyCode)
    keyRelease(KeyEvent.VK_META)
}

/**
 * Simulates pressing CMD + doing something else.
 *
 * @param action The function to be done in combination with the CMD key.
 */
fun Robot.command(action: (Robot) -> Unit) {
    keyPress(KeyEvent.VK_META)
    action(this)
    keyRelease(KeyEvent.VK_META)
}

/**
 * Simulates pressing SHIFT + another key.
 *
 * @param keyCode The integer code of the key to be pressed in combination with the SHIFT key.
 */
fun Robot.shift(keyCode: Int) {
    keyPress(KeyEvent.VK_SHIFT)
    type(keyCode)
    keyRelease(KeyEvent.VK_SHIFT)
}

/**
 * Simulates pressing SHIFT + doing something else.
 *
 * @param action The function to be done in combination with the SHIFT key.
 */
fun Robot.shift(action: (Robot) -> Unit) {
    keyPress(KeyEvent.VK_SHIFT)
    action(this)
    keyRelease(KeyEvent.VK_SHIFT)
}

/**
 * Simulates pressing WINDOWS + doing something else.
 *
 * @param action The function to be done in combination with the WINDOWS key.
 */
fun Robot.windows(action: (Robot) -> Unit) {
    keyPress(KeyEvent.VK_WINDOWS)
    action(this)
    keyRelease(KeyEvent.VK_WINDOWS)
}

/**
 * Simulates pressing ALT + doing something else.
 *
 * @param action The function to be done in combination with the ALT key.
 */
fun Robot.alt(action: (Robot) -> Unit) {
    keyPress(KeyEvent.VK_ALT)
    action(this)
    keyRelease(KeyEvent.VK_ALT)
}

/**
 * Scrolls the mouse wheel up or down based on the specified direction.
 *
 * @param direction A string specifying the scroll direction ("up" or "down").
 */
fun Robot.scroll(direction: String) {
    direction.split(" ").forEach { dir ->
        when (dir.lowercase()) {
            "up" -> mouseWheel(-1)
            "down" -> mouseWheel(1)
        }
    }
}

/**
 * Moves the mouse cursor based on a string of directions.
 *
 * @param directions A string containing directions (e.g., "up down left right") to move the mouse
 *   cursor. Each direction should be separated by a space.
 */
fun Robot.arrow(directions: String) {
    directions.split(" ").forEach { direction ->
        when (direction.lowercase()) {
            "up" -> type(KeyEvent.VK_UP)
            "down" -> type(KeyEvent.VK_DOWN)
            "left" -> type(KeyEvent.VK_LEFT)
            "right" -> type(KeyEvent.VK_RIGHT)
            "write" -> type(KeyEvent.VK_RIGHT)
        }
    }
}

/** Simulates saving something by pressing CONTROL + S. */
fun Robot.save() = control(VK_S)

/**
 * Helper function to simulate mouse clicks.
 *
 * @param mask The mask for the mouse button to be clicked.
 */
private fun Robot.click(mask: Int) {
    mousePress(mask)
    mouseRelease(mask)
}
